flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReader.java

```diff
@@ -0,0 +1,59 @@
+package org.apache.flink.runtime.checkpoint.channel;
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+
+/**
+ * Reads channel state saved during checkpoint/savepoint.
+ */
+@Internal
+public interface ChannelStateReader extends AutoCloseable {
+
+	/**
+	 * ReadResult.
+	 */
+	enum ReadResult {HAS_MORE_DATA, NO_MORE_DATA}
+
+	/**
+	 * Put data into the supplied buffer to be injected into {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 */
+	ReadResult readInputData(InputChannelInfo info, Buffer buffer);
+
+	/**
+	 * Put data into the supplied buffer to be injected into {@link org.apache.flink.runtime.io.network.partition.ResultSubpartition ResultSubpartition}.
+	 */
+	ReadResult readOutputData(ResultSubpartitionInfo info, Buffer buffer);
+
+	@Override
+	default void close() throws Exception {
```

Comment:
How many implementation with no-op close do you expect? I'd assume it's only one and then I'd avoid having it as default implementation.